client/network/src/service/out_events.rs

```diff
@@ -0,0 +1,267 @@
+			Some(Metrics::register(registry)?)
+		} else {
+			None
+		};
+
+		Ok(OutChannels {
+			event_streams: Vec::new(),
+			metrics: Arc::new(metrics),
+		})
+	}
+
+	/// Adds a new [`Sender`] to the collection.
+	pub fn push(&mut self, sender: Sender) {
+		let mut metrics = sender.metrics.lock();
+		debug_assert!(metrics.is_none());
+		*metrics = Some(self.metrics.clone());
+		drop(metrics);
+		self.event_streams.push(sender);
+
+		if let Some(metrics) = &*self.metrics {
+			metrics.out_events_num_channels.inc();
+		}
+	}
+
+	/// Sends an event.
+	pub fn send(&mut self, event: Event) {
+		self.event_streams.retain(|sender| {
+			sender.inner.unbounded_send(event.clone()).is_ok()
+		});
+
+		// The number of elements remaining in `event_streams` corresponds to the number of times
+		// we have sent an element on the channel.
+		if let Some(metrics) = &*self.metrics {
+			increase_metrics(metrics, &event, self.event_streams.len() as u64);
+		}
+	}
+}
+
+impl fmt::Debug for OutChannels {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_struct("OutChannels")
+			.field("num_channels", &self.event_streams.len())
+			.finish()
+	}
+}
+
+struct Metrics {
+	// This list is ordered alphabetically
+	out_events_dht_count: Gauge<U64>,
+	out_events_num_channels: Gauge<U64>,
+	out_events_notifications_closed_count: GaugeVec<U64>,
+	out_events_notifications_opened_count: GaugeVec<U64>,
+	out_events_notifications_sizes: GaugeVec<U64>,
+}
+
+impl Metrics {
+	fn register(registry: &Registry) -> Result<Self, PrometheusError> {
+		Ok(Self {
+			out_events_dht_count: register(Gauge::new(
+				"sub_libp2p_out_events_dht_count",
+				"Number of DHT events currently pending in the channels that broadcast network events",
+			)?, registry)?,
+			out_events_num_channels: register(Gauge::new(
+				"sub_libp2p_out_events_num_channels",
+				"Number of internal active channels that broadcast network events",
+			)?, registry)?,
+			out_events_notifications_closed_count: register(GaugeVec::new(
+				Opts::new(
+					"sub_libp2p_out_events_notifications_closed_count",
+					"Number of notification substreams opened events pending in the channels that broadcast network events"
+				),
+				&["protocol"]
+			)?, registry)?,
+			out_events_notifications_opened_count: register(GaugeVec::new(
+				Opts::new(
+					"sub_libp2p_out_events_notifications_opened_count",
+					"Number of notification substreams opened events pending in the channels that broadcast network events"
+				),
+				&["protocol"]
+			)?, registry)?,
+			out_events_notifications_sizes: register(GaugeVec::new(
+				Opts::new(
+					"sub_libp2p_out_events_notifications_sizes",
+					"Total size of notification events pending in the channels that broadcast network events"
+				),
+				&["protocol"]
+			)?, registry)?,
+		})
+	}
+}
+
+fn increase_metrics(metrics: &Metrics, event: &Event, num: u64) {
+	match event {
+		Event::Dht(_) => metrics.out_events_dht_count.inc(),
```

Comment:
```suggestion
		Event::Dht(_) => metrics.out_events_dht_count.add(num),
```

Shouldn't this also be increased by the amount of channels we duplicated the event to?